### PR TITLE
bpo-37961: tracemalloc: store the actual length of traceback

### PR DESCRIPTION
**Pull request by jd**
[bpo-37961](https://bugs.python.org/issue37961): tracemalloc: store the actual length of traceback

This adds a new field to the traceback_t data structure so it stores the
original length of the traceback that was recorded. This is useful to know if
the traceback has been truncated or not.

https://bugs.python.org/issue37961

### CODE REVIEW

File: Modules/_tracemalloc.c [hunk 1]
```diff
@@ -436,11 +443,12 @@ traceback_get_frames(traceback_t *traceback)
...
-        assert(traceback->frames[traceback->nframe].filename != NULL);
-        traceback->nframe++;
-        if (traceback->nframe == _Py_tracemalloc_config.max_nframe)
-            break;
+        if (traceback->nframe < _Py_tracemalloc_config.max_nframe) {
+            tracemalloc_get_frame(pyframe, &traceback->frames[traceback->nframe]);
+            assert(traceback->frames[traceback->nframe].filename != NULL);
+            traceback->nframe++;
+        }
+        traceback->length++;
```
Comment (vstinner) [inline Modules/_tracemalloc.c]: You should handle integer overflow: stop if INT_MAX is reached. If should be unlikely, but I prefer to be safe.

File: Modules/_tracemalloc.c [hunk 2]
```diff
@@ -78,7 +78,10 @@ __attribute__((packed))
 
 typedef struct {
     Py_uhash_t hash;
+    /* Number of frames stored */
     int nframe;
+    /* Total number of frames the traceback had */
+    int length;
```
Comment (vstinner) [inline Modules/_tracemalloc.c]: Hum, maybe rename it to "total" or "total_nframe"?

File: Modules/_tracemalloc.c [hunk 3]
```diff
@@ -1259,6 +1269,13 @@ trace_to_pyobject(unsigned int domain, trace_t *trace,
     }
     PyTuple_SET_ITEM(trace_obj, 2, obj);
 
+    obj = PyLong_FromSize_t(trace->traceback->length);
```
Comment (vstinner) [inline Modules/_tracemalloc.c]: nitpick: I would prefer to use PyLong_FromLong() which is more common and closer to int type.

File: Lib/tracemalloc.py [hunk 4]
```diff
@@ -282,6 +282,13 @@ def size(self):
     def traceback(self):
         return Traceback(self._trace[2])
 
+    @property
+    def length(self):
```
Comment (vstinner) [inline Lib/tracemalloc.py]: I suggest to rename it total_nframe and move it to Traceback.

**Review by vstinner — CHANGES_REQUESTED** [review]
You should also update Doc/library/tracemalloc.rst documentation, document Trace.length new attribute (I suggest to add Traceback.total_nframe attribute instead):
https://docs.python.org/dev/library/tracemalloc.html#trace

Is it possible to load a pickle file created by Python 3.7 on Python 3.9 with this change?

File: Modules/_tracemalloc.c (hunk 1 above)
Comment (jd) [reply Modules/_tracemalloc.c]: Good point!

File: Modules/_tracemalloc.c (hunk 2 above)
Comment (jd) [reply Modules/_tracemalloc.c]: Let's go for `total_nframe`.

File: Lib/tracemalloc.py (hunk 4 above)
Comment (jd) [reply Lib/tracemalloc.py]: Ok!

File: Modules/_tracemalloc.c (hunk 3 above)
Comment (jd) [reply Modules/_tracemalloc.c]: ok!

**Comment by jd** [comment]
> Is it possible to load a pickle file created by Python 3.7 on Python 3.9 with this change?

No, it failed with:
```
>>> s.traces[0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jd/Source/cpython/Lib/tracemalloc.py", line 299, in __repr__
    % (self.domain, _format_size(self.size, False), self.traceback))
  File "/Users/jd/Source/cpython/Lib/tracemalloc.py", line 284, in traceback
    return Traceback(self._trace[2], self._trace[3])
IndexError: tuple index out of range
```

I fixed it by setting a default `nframe` to `None` if the info is not available.

**Comment by jd** [comment]
@vstinner I just saw your comment on the issue. I'll update this to also reduce the size of the already existing `nframe` attribute.

File: Modules/_tracemalloc.c [hunk 5]
```diff
@@ -78,15 +78,18 @@ __attribute__((packed))
...
+    uint16_t total_nframe;
     frame_t frames[1];
 } traceback_t;
 
 #define TRACEBACK_SIZE(NFRAME) \
         (sizeof(traceback_t) + sizeof(frame_t) * (NFRAME - 1))
 
 #define MAX_NFRAME \
-        ((INT_MAX - (int)sizeof(traceback_t)) / (int)sizeof(frame_t) + 1)
+        ((UINT16_MAX - (int)sizeof(traceback_t)) / (int)sizeof(frame_t) + 1)
```
Comment (jd) [inline Modules/_tracemalloc.c]: @vstinner I'm not really sure about this change. I actually can't understand how `MAX_NFRAME` is calculated here. Can you give me a hint?

Comment (vstinner) [reply Modules/_tracemalloc.c]: It looks like I wrote to ensure that TRACEBACK_SIZE(nframe) will not overflow, but I am not sure that using INT_MAX makes sense.

TRACEBACK_SIZE() maximum should be smaller than or equal to SIZE_MAX, not INT_MAX.

Maybe Py_MIN(INT_MAX, ((SIZE_MAX - sizeof(traceback_t)) / sizeof(frame_t) + 1) would be more correct?

Comment (jd) [reply Modules/_tracemalloc.c]: > TRACEBACK_SIZE() maximum should be smaller than or equal to SIZE_MAX, not INT_MAX.

Agreed. `TRACEBACK_SIZE()` is actually a `size_t`.

> Py_MIN(INT_MAX, ((SIZE_MAX - sizeof(traceback_t)) / sizeof(frame_t) + 1) would be more correct?

Yes. I think you either want the maximum size you can allocate or the maximum number of frame you can store in a `uint16_t`.

I'll change that.

It'll make the result 65535 in all cases for now :)

**Comment by jd** [comment]
That should be ready for review but I see @bedevere-bot might be confused with labels. Ping @vstinner?

File: Lib/tracemalloc.py [hunk 6]
```diff
@@ -378,7 +382,7 @@ def _match_traceback(self, traceback):
             return self._match_frame(filename, lineno)
 
     def _match(self, trace):
-        domain, size, traceback = trace
+        domain, size, traceback, length = trace
```
Comment (vstinner) [inline Lib/tracemalloc.py]: I understand that this line breaks if loading from a Python 3.8. Maybe you should add the 4th field in _Traces constructor, if len(trace) == 3, rather than fixing it in the traceback property.

For example:

```
# Convert Python 3.8 trace: add nframe field (set to None)
if traces and len(traces[0]) == 3:
   traces = tuple(trace + (None,) for trace in traces)
```

File: Modules/_tracemalloc.c [hunk 7]
```diff
@@ -78,15 +78,21 @@ __attribute__((packed))
...
 
 #define TRACEBACK_SIZE(NFRAME) \
         (sizeof(traceback_t) + sizeof(frame_t) * (NFRAME - 1))
 
+/* The maximum number of frames is either:
+ - The maximum number of frames we can store in `nframe`
+ - The maximum memory size_t we can allocate */
 #define MAX_NFRAME \
-        ((INT_MAX - (int)sizeof(traceback_t)) / (int)sizeof(frame_t) + 1)
+        Py_MIN(UINT16_MAX, ((SIZE_MAX - sizeof(traceback_t)) / sizeof(frame_t) + 1))
```
Comment (vstinner) [inline Modules/_tracemalloc.c]: Can you try to convert to macro to a constant variable? Otherwise, the macro is expended each time it is used, and the compiler might recompute it each time.

```
static const int MAX_NFRAME = Py_MIN(UINT16_MAX, ((SIZE_MAX - sizeof(traceback_t)) / sizeof(frame_t) + 1));
```

I suggest to use int (signed). If you want to use an unsigned int, you need to modify more change.

You may need to update code around `if (max_nframe < 1 || (unsigned long) max_nframe > MAX_NFRAME) {`.

File: Modules/_tracemalloc.c [hunk 8]
```diff
@@ -78,15 +78,21 @@ __attribute__((packed))
...
+    /* Total number of frames the traceback had */
+    uint16_t total_nframe;
     frame_t frames[1];
 } traceback_t;
 
 #define TRACEBACK_SIZE(NFRAME) \
         (sizeof(traceback_t) + sizeof(frame_t) * (NFRAME - 1))
 
+/* The maximum number of frames is either:
+ - The maximum number of frames we can store in `nframe`
```
Comment (vstinner) [inline Modules/_tracemalloc.c]:
```suggestion
 - The maximum number of frames we can store in `traceback_t.nframe`
```

File: Lib/tracemalloc.py [hunk 9]
```diff
@@ -182,15 +182,16 @@ class Traceback(Sequence):
...
 
-    def __init__(self, frames):
+    def __init__(self, frames, nframe):
         Sequence.__init__(self)
         # frames is a tuple of frame tuples: see Frame constructor for the
         # format of a frame tuple; it is reversed, because _tracemalloc
         # returns frames sorted from most recent to oldest, but the
         # Python API expects oldest to most recent
         self._frames = tuple(reversed(frames))
+        self.nframe = nframe
```
Comment (vstinner) [inline Lib/tracemalloc.py]: I would prefer to make it private and add a read-only property. The API is designed to make all data read-only to ensure that all data remain consistent.

File: Doc/library/tracemalloc.rst [hunk 10]
```diff
@@ -659,12 +659,19 @@ Traceback
 
    When a snapshot is taken, tracebacks of traces are limited to
    :func:`get_traceback_limit` frames. See the :func:`take_snapshot` function.
+   The original number of frames of the traceback is stored in the
+   :attr:`Traceback.nframe` attribute. That allows to know if a traceback has
+   been truncated by the traceback limit.
 
    The :attr:`Trace.traceback` attribute is an instance of :class:`Traceback`
    instance.
```
Comment (vstinner) [inline Doc/library/tracemalloc.rst:670]: You should declare the new field using ".. attribute:: " markup.

**Review by vstinner — COMMENTED** [review]
In C, you chose "total_nframe" name. In Python, it's "nframe" in the API, and sometimes "length" in the code. I would prefer to always use the same name everywhere.

I like "total_nframe", but I don't have a strong preference :-)

File: Lib/tracemalloc.py (hunk 9 above)
Comment (jd) [reply Lib/tracemalloc.py]: Sure!

File: Modules/_tracemalloc.c (hunk 7 above)
Comment (jd) [reply Modules/_tracemalloc.c]: Not sure it's worth it, but fine. AFAIU the macro is expanded only when pre-processing `_tracemalloc.` which should only happen once.

The result is an `unsigned long`, no matter what. We can cast it to something else, but that should be ok — the code compiles without warning AFAICT.

**Comment by jd** [comment]
@vstinner Updated with your comments :)

File: Lib/tracemalloc.py [hunk 11]
```diff
@@ -280,7 +287,10 @@ def size(self):
 
     @property
     def traceback(self):
-        return Traceback(self._trace[2])
+        # Python < 3.9
+        if len(self._trace) == 3:
+            return Traceback(self._trace[2], None)
```
Comment (vstinner) [inline Lib/tracemalloc.py]: This code path becomes deadcode with "traces = tuple(trace + (None,) for trace in traces)" below, no? I guess that you can now remove this code.

File: Modules/_tracemalloc.c [hunk 12]
```diff
@@ -436,11 +445,13 @@ traceback_get_frames(traceback_t *traceback)
...
-        assert(traceback->frames[traceback->nframe].filename != NULL);
-        traceback->nframe++;
-        if (traceback->nframe == _Py_tracemalloc_config.max_nframe)
-            break;
+        if (traceback->nframe < _Py_tracemalloc_config.max_nframe) {
+            tracemalloc_get_frame(pyframe, &traceback->frames[traceback->nframe]);
+            assert(traceback->frames[traceback->nframe].filename != NULL);
+            traceback->nframe++;
+        }
+        if (traceback->total_nframe++ < UINT16_MAX)
```
Comment (vstinner) [inline Modules/_tracemalloc.c]: Hum, I don't think that this code properly handles integer overlow, since traceback->total_nframe++ has a side effect. It should be:

```suggestion
        if (traceback->total_nframe < UINT16_MAX)
```

File: Modules/_tracemalloc.c [hunk 13]
```diff
@@ -1046,10 +1059,10 @@ tracemalloc_start(int max_nframe)
     PyMemAllocatorEx alloc;
     size_t size;
 
-    if (max_nframe < 1 || max_nframe > MAX_NFRAME) {
+    if (max_nframe < 1 || (unsigned long) max_nframe > MAX_NFRAME) {
         PyErr_Format(PyExc_ValueError,
-                     "the number of frames must be in range [1; %i]",
-                     (int)MAX_NFRAME);
+                     "the number of frames must be in range [1; %u]",
```
Comment (vstinner) [inline Modules/_tracemalloc.c]: unsigned long requires `%lu` format:

```suggestion
                     "the number of frames must be in range [1; %lu]",
```

**Review by vstinner — COMMENTED** [review]
Thanks, the updated PR looks better. Another round of minor comments.

You may mention total_nfrace in start() documentation: https://docs.python.org/dev/library/tracemalloc.html#tracemalloc.start Users may want to use it to adjust the next call to start().

File: Lib/tracemalloc.py (hunk 11 above)
Comment (jd) [reply Lib/tracemalloc.py]: You're right! I forgot we added that.

File: Modules/_tracemalloc.c (hunk 12 above)
Comment (jd) [reply Modules/_tracemalloc.c]: Woh good catch, this is clearly a bad typo.

**Comment by jd** [comment]
Here you go @vstinner :)

File: Lib/tracemalloc.py [hunk 14]
```diff
@@ -182,15 +182,20 @@ class Traceback(Sequence):
     Sequence of Frame instances sorted from the oldest frame
     to the most recent frame.
     """
-    __slots__ = ("_frames",)
+    __slots__ = ("_frames", '_total_nframe')
 
-    def __init__(self, frames):
+    def __init__(self, frames, total_nframe):
```
Comment (vstinner) [inline Lib/tracemalloc.py]: What do you think of making this new parameter optional (total_nframe=None), and use len(frames) by default, if it's not specified? It would simplify get_object_traceback() and test_tracemalloc.

File: Modules/_tracemalloc.c [hunk 15]
```diff
@@ -1259,6 +1271,13 @@ trace_to_pyobject(unsigned int domain, trace_t *trace,
     }
     PyTuple_SET_ITEM(trace_obj, 2, obj);
 
+    obj = PyLong_FromLong(trace->traceback->total_nframe);
```
Comment (vstinner) [inline Modules/_tracemalloc.c]: Since it's unsigned, I suggest to use PyLong_FromUnsignedLong() to avoid the risk of introducing a compiler warning.

**Review by vstinner — COMMENTED** [review]
The overall change now LGTM, but I have one suggestion about the tracemalloc.Traceback constructor which may simplify the code and leaves the API kind of backward compatible.

File: Modules/_tracemalloc.c (hunk 15 above)
Comment (jd) [reply Modules/_tracemalloc.c]: Good catch. Done.

File: Lib/tracemalloc.py (hunk 14 above)
Comment (jd) [reply Lib/tracemalloc.py]: That means you lose information when you load traces from Python <= 3.8 as there's no way to know if the `total_nframe` was captured or simply faked at load time.
I think it's better to say that you either that information or you don't (so set to `None`) rather than trying to make the field ambiguous.

**Comment by jd** [comment]
Last update I hope 🤞 @vstinner :)

**Comment by jd** [comment]
> I'm not sure that I understand your comment. Do you suggest to modify your PR to allow total_nframe to be None? I think that I'm fine with that.

You were suggesting to always set `total_nframe` to some value — either the one provided in the argument list, or, if set to `None`, to compute it with `len(frames)`. I'm just saying that doing so would make `None` an invalid value for `total_nframe` which removes a piece of information that might be important. :)

> _group_by() and loading a Python 3.8 file would use total_nframe=None if I understand correctly. In that case, I suggest to modify Traceback.repr() to omit "total_nframe=%s" if it's None.
> By the way, "<Traceback %r total_nframe=%s>" may be better to first show the most important first (the frames). So "<Traceback %r total_nframe=%s>" or "<Traceback %r>" depending if total_nframe is None.

Why not, done.

> In your implementation, total_nframe is lost in tracemalloc.get_object_traceback(). I'm not sure if it's a deliberate choice or not.

No, it's not. It's just a "feature not implemented yet". :)

> If you want, I can write this complex change, once this PR lands.

Yeah, I think it'd be safer to go in a new PR that leverage this one rather than making this one heavier. 👍

**Comment by vstinner** [comment]
> You were suggesting to always set total_nframe to some value — either the one provided in the argument list, or, if set to None, to compute it with len(frames).

In your PR, total_nframe is always set to an integer. When you don't know the value, you put len(frames), but you do that in the Traceback caller. I suggest to do that in the constructor to simplify your PR: tests, get_object_traceback(), _group_by(), etc. would just call Traceback(frames).

> I'm just saying that doing so would make None an invalid value for total_nframe which removes a piece of information that might be important. :)

Again, your PR always set total_nframe to a number, so I don't understand your remark. Do you plan to modify your PR to set total_nframe to None in some cases (ex: _group_by())? If yes, you should update the documentation to mention that total_nframe can be None.

**Comment by jd** [comment]
> In your PR, total_nframe is always set to an integer. When you don't know the value, you put len(frames), but you do that in the Traceback caller. I suggest to do that in the constructor to simplify your PR: tests, get_object_traceback(), _group_by(), etc. would just call Traceback(frames).

That's mainly because the argument *was* mandatory in an earlier version IIRC. I can just remove it from the tests now if you prefer.

> Again, your PR always set total_nframe to a number, so I don't understand your remark. Do you plan to modify your PR to set total_nframe to None in some cases (ex: _group_by())? If yes, you should update the documentation to mention that total_nframe can be None.

The initial plan was to always have it set to the number of frames — remember, it was named `nframes`. So it would either be the number of frames we captured or — if available — the number of total frames that were originally in place. So `None` was not possible value.

Now, the way the PR has changed, `None` could be a value. I've updated the PR in that sense. Some tests now uses `None` has a value. I've updated the attribute doc.

File: Lib/tracemalloc.py [hunk 16]
```diff
@@ -221,7 +226,12 @@ def __str__(self):
         return str(self[0])
 
     def __repr__(self):
-        return "<Traceback %r>" % (tuple(self),)
+        s = "<Traceback %r" % tuple(self)
+        if self._total_nframe is None:
+            s += ">"
+        else:
+            s += "total_nframe=%d>" % self.total_nframe
```
Comment (vstinner) [inline Lib/tracemalloc.py]: You forgot a space. I like f-string :-) No need to be pedantic with %d.

```suggestion
            s += f" total_nframe={self.total_nframe}>"
```

File: Lib/tracemalloc.py [hunk 17]
```diff
@@ -247,7 +257,8 @@ def format(self, limit=None, most_recent_first=False):
 def get_object_traceback(obj):
     """
     Get the traceback where the Python object *obj* was allocated.
-    Return a Traceback instance.
+    Return a Traceback instance
+    .
```
Comment (vstinner) [inline Lib/tracemalloc.py]: Is it a deliberate chose to put the dot on a newline? If not, please revert this change.

File: Lib/tracemalloc.py [hunk 18]
```diff
@@ -280,7 +291,7 @@ def size(self):
 
     @property
     def traceback(self):
-        return Traceback(self._trace[2])
+        return Traceback(self._trace[2], self._trace[3])
```
Comment (vstinner) [inline Lib/tracemalloc.py]: Since total_nframe can be None, maybe now it would make more sense to handle Python 3.8 compatibility on demand in this property:

```
# Call Traceback(frames, total_nframe).
# For a Python 3.8 trace, call Traceback(frames).
return Traceback(*self._trace[2:])
```

(and revert change in `_Traces.__init__`)

Well, you wrote something like that in a previous PR :-)

**Review by vstinner — COMMENTED** [review]
> The initial plan was to always have it set to the number of frames — remember, it was named nframes. So it would either be the number of frames we captured or — if available — the number of total frames that were originally in place. So None was not possible value.
> Now, the way the PR has changed, None could be a value. I've updated the PR in that sense. Some tests now uses None has a value. I've updated the attribute doc.

Oh, I see, the PR evolved and then was longer consistent. I'm fine with having total_nframe=None. It's just that it wasn't clear if it could be None or not. It's better with your update PR.

New review.

File: Lib/tracemalloc.py (hunk 16 above)
Comment (jd) [reply Lib/tracemalloc.py]: Thanks!

File: Lib/tracemalloc.py (hunk 18 above)
Comment (jd) [reply Lib/tracemalloc.py]: Good idea!

File: Lib/tracemalloc.py (hunk 17 above)
Comment (jd) [reply Lib/tracemalloc.py]: No sorry, PEBKAC.

File: Doc/library/tracemalloc.rst [hunk 19]
```diff
@@ -659,13 +662,24 @@ Traceback
...
 
    .. versionchanged:: 3.7
       Frames are now sorted from the oldest to the most recent, instead of most recent to oldest.
 
+   .. attribute:: total_nframe
+
+      Total number of frames that composed the traceback before truncation.
+      This attribute can be set to `None` if the information is not available.
```
Comment (vstinner) [inline Doc/library/tracemalloc.rst]: Travis CI fails with:

```
[2] library/tracemalloc.rst:678: default role used
````

I guess that it should be:

```suggestion
      This attribute can be set to ``None`` if the information is not available.
```

**Review by vstinner — APPROVED** [review]
LGTM. But the CI blocks the PR.

**Comment by jd** [comment]
Thanks, fixed the doc!

**Comment by vstinner** [comment]
> https://buildbot.python.org/all/#builders/103/builds/3247

That's https://bugs.python.org/issue35998
